src/file_path.rs

```diff
@@ -0,0 +1,160 @@
+#[derive(PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[repr(transparent)]
+pub struct FilePath(#[cfg(windows)] str, #[cfg(not(windows))] [u8]);
+
+impl FilePath {
+    pub fn new<P: AsRef<FilePath>>(p: &P) -> &Self {
+        p.as_ref()
+    }
+
+    pub fn as_bytes(&self) -> &[u8] {
+        self.as_ref()
+    }
+}
+
+impl Debug for FilePath {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        f.write_char('"')?;
+
+        #[cfg(windows)]
+        write!(f, "{}", self.0.escape_debug())?;
+
+        #[cfg(not(windows))]
+        for chunk in self.0.utf8_chunks() {
+            write!(f, "{}", chunk.valid().escape_debug())?;
+            write!(f, "{}", chunk.invalid().escape_ascii())?;
+        }
+
+        f.write_char('"')
+    }
+}
+
+impl AsRef<FilePath> for str {
+    fn as_ref(&self) -> &FilePath {
+        #[cfg(windows)]
+        // SAFETY: On windows FilePath is a str. As `self` is a str as well
+        // and FilePath is repr(transparent) this is sound.
+        unsafe {
+            transmute(self)
+        }
+
+        #[cfg(not(windows))]
+        self.as_bytes().as_ref()
+    }
+}
+
+impl AsRef<FilePath> for String {
+    fn as_ref(&self) -> &FilePath {
+        self.as_str().as_ref()
+    }
+}
+
+#[cfg(not(windows))]
+impl AsRef<FilePath> for [u8] {
+    fn as_ref(&self) -> &FilePath {
+        // SAFETY: On non-windows FilePath is a byte slice. As `self` is a byte slice as well
+        // and FilePath is repr(transparent) this is sound.
+        unsafe { transmute(self) }
+    }
+}
+
+#[cfg(any(unix, target_os = "wasi"))]
+impl AsRef<FilePath> for OsStr {
+    fn as_ref(&self) -> &FilePath {
+        self.as_bytes().as_ref()
+    }
+}
+
+#[cfg(any(unix, target_os = "wasi"))]
+impl AsRef<FilePath> for Path {
+    fn as_ref(&self) -> &FilePath {
+        self.as_os_str().as_ref()
+    }
+}
+
+impl AsRef<[u8]> for FilePath {
+    fn as_ref(&self) -> &[u8] {
+        #[cfg(windows)]
+        {
+            self.0.as_ref()
+        }
+
+        #[cfg(not(windows))]
+        {
+            &self.0
+        }
+    }
+}
+
+#[cfg(any(windows, unix, target_os = "wasi"))]
+impl AsRef<OsStr> for FilePath {
+    fn as_ref(&self) -> &OsStr {
+        #[cfg(windows)]
+        {
+            self.0.as_ref()
+        }
+
+        #[cfg(any(unix, target_os = "wasi"))]
+        {
+            OsStr::from_bytes(&self.0)
+        }
+    }
+}
+
+#[cfg(any(windows, unix, target_os = "wasi"))]
+impl AsRef<Path> for FilePath {
+    fn as_ref(&self) -> &Path {
+        Path::new(self)
+    }
+}
+
+impl<'a> TryFrom<&'a OsStr> for &'a FilePath {
+    type Error = Error;
+    fn try_from(value: &'a OsStr) -> Result<Self, Self::Error> {
+        #[cfg(not(any(unix, target_os = "wasi")))]
+        {
+            Ok(value.to_str().ok_or(Error::InvalidUtf8)?.as_ref())
+        }
+
+        #[cfg(any(unix, target_os = "wasi"))]
+        {
+            Ok(value.as_ref())
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
```

Comment:
Personally I feel like a few more tests would be nice - e.g. validating the `TryFrom<&'a OsStr> for &'a FilePath` impl.

Reply:
Thx for the callout, wouldn't've realized that I was missing the `?Sized` bound on `FilePath::new`